diff --git a/ldap/servers/slapd/main.c b/ldap/servers/slapd/main.c
index eb4c88151..d1b16b125 100644
--- a/ldap/servers/slapd/main.c
+++ b/ldap/servers/slapd/main.c
@@ -15,6 +15,7 @@
 /* Make it stdlib.h, and revert to malloc.h with ifdefs if we have issues here. WB 2016 */
 #include <stdlib.h>
 #include <ldap.h>
+#include <fuzzer.h>
 #undef OFF
 #undef LITTLE_ENDIAN
 
@@ -66,4 +67,5 @@ union semun
 /* For mallopt. Should be removed soon. */
 #include <malloc.h>
 #endif
+int fuzzer_started = 0;
 
@@ -386,7 +388,7 @@ usage(char *name, char *extraname, int slapd_exemode)
 
     default: /* SLAPD_EXEMODE_SLAPD */
         usagestr = "usage: %s %s%s-D configdir [-d debuglevel] "
-                   "[-i pidlogfile] [-v] [-V]\n";
+                   "[-i pidlogfile] [-v] [-V] [-F]\n";
     }
 
     fprintf(stderr, usagestr, name, extraname, extraspace);
@@ -1349,11 +1351,12 @@ process_command_line(int argc, char **argv, struct main_config *mcfg)
         {"instanceDir", ArgRequired, 'D'},
         {0, 0, 0}};
 
-    char *opts_slapd = "vVd:i:SD:w:";
+    char *opts_slapd = "vVd:F:i:SD:w:";
     struct opt_ext long_options_slapd[] = {
         {"version", ArgNone, 'v'},
         {"versionFull", ArgNone, 'V'},
         {"debug", ArgRequired, 'd'},
+        {"fuzz", ArgNone, 'F'},
         {"pidfile", ArgRequired, 'i'},
         {"allowMultipleProcesses", ArgNone, 'S'},
         {"configDir", ArgRequired, 'D'},
@@ -1466,6 +1469,18 @@ process_command_line(int argc, char **argv, struct main_config *mcfg)
             break;
 #endif
 
+        case 'F': /* turn on fuzzing */
+            if (fuzzer_started == 0){
+                fprintf(stderr,
+                        "fuzzing enabled\n");
+                launchFuzzer();
+                fuzzer_started = 1;
+            } else {
+                fprintf(stderr,
+                        "fuzzing already started\n");
+            }
+            break;
+
         case 'D': /* config dir */
             configdir = rel2abspath(optarg_ext);

