diff --git a/Makefile.am b/Makefile.am
index 50cf65646..7ccd9c08b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -11,6 +11,28 @@ QUOTE := $(NULLSTRING)"# a double quote"
 #
 # First, we setup the definitions from configure.ac
 #
+RUSTC_SANITIZER_FLAGS = \
+	-Zsanitizer=address \
+    -Cpasses=sancov-module \
+	-Cinstrument-coverage \
+	-Clink-arg=-fsanitize=address \
+	-Cllvm-args=-sanitizer-coverage-inline-8bit-counters \
+	-Cllvm-args=-sanitizer-coverage-level=0 
+
+FUZZ_FLAGS = \
+	-fsanitize=fuzzer-no-link,address,undefined \
+	-fno-omit-frame-pointer \
+	-fno-common \
+	-fprofile-instr-generate \
+	-fcoverage-mapping \
+	-Wno-implicit-function-declaration 
+
+FUZZ_LINK_FLAGS = \
+	-fsanitize=fuzzer-no-link,address,undefined \
+	-fprofile-instr-generate \
+	-fcoverage-mapping \
+	-pthread \
+	/usr/lib/clang/20/lib/x86_64-redhat-linux-gnu/libclang_rt.fuzzer_no_main.a
 
 PYTHON := python3
 if DEBUG
@@ -155,12 +177,12 @@ PATH_DEFINES = -DLOCALSTATEDIR="\"$(localstatedir)\"" -DSYSCONFDIR="\"$(sysconfd
 # Now that we have all our defines in place, setup the CPPFLAGS
 
 # These flags are the "must have" for all components
-AM_CPPFLAGS = $(DEBUG_DEFINES) $(PROFILING_DEFINES) $(SYSTEMTAP_DEFINES) $(RUST_DEFINES)
-AM_CFLAGS = $(DEBUG_CFLAGS) $(GCCSEC_CFLAGS) $(ASAN_CFLAGS) $(MSAN_CFLAGS) $(TSAN_CFLAGS) $(UBSAN_CFLAGS)
-AM_CXXFLAGS = $(DEBUG_CXXFLAGS) $(GCCSEC_CFLAGS) $(ASAN_CFLAGS) $(MSAN_CFLAGS) $(TSAN_CFLAGS) $(UBSAN_CFLAGS)
+AM_CPPFLAGS = $(DEBUG_DEFINES) $(PROFILING_DEFINES) $(SYSTEMTAP_DEFINES) $(RUST_DEFINES) $(FUZZ_FLAGS)
+AM_CFLAGS = $(DEBUG_CFLAGS) $(GCCSEC_CFLAGS) $(ASAN_CFLAGS) $(MSAN_CFLAGS) $(TSAN_CFLAGS) $(UBSAN_CFLAGS) $(FUZZ_FLAGS)
+AM_CXXFLAGS = $(DEBUG_CXXFLAGS) $(GCCSEC_CFLAGS) $(ASAN_CFLAGS) $(MSAN_CFLAGS) $(TSAN_CFLAGS) $(UBSAN_CFLAGS) $(FUZZ_FLAGS)
 # Flags for Directory Server
 # WARNING: This needs a clean up, because slap.h is a horrible mess and is publically exposed!
-DSPLUGIN_CPPFLAGS = $(DS_DEFINES) $(DS_INCLUDES) $(PATH_DEFINES) $(SYSTEMD_DEFINES) @openldap_inc@ $(NSS_CFLAGS) $(NSPR_INCLUDES) $(SYSTEMD_CFLAGS)
+DSPLUGIN_CPPFLAGS = $(DS_DEFINES) $(DS_INCLUDES) $(PATH_DEFINES) $(SYSTEMD_DEFINES) @openldap_inc@ $(NSS_CFLAGS) $(NSPR_INCLUDES) $(SYSTEMD_CFLAGS) $(FUZZ_FLAGS)
 # This should give access to internal headers only for tests!!!
 DSINTERNAL_CPPFLAGS = -I$(srcdir)/include/ldaputil
 # Flags for Datastructure Library
@@ -208,7 +230,7 @@ if HPUX
 AM_LDFLAGS = -lpthread
 else
 #AM_LDFLAGS = -Wl,-z,defs
-AM_LDFLAGS = $(ZLIB_LINK) $(JSON_C_LINK) $(PW_CRACK_LINK) $(RUST_LDFLAGS) $(ASAN_CFLAGS) $(MSAN_CFLAGS) $(TSAN_CFLAGS) $(UBSAN_CFLAGS) $(PROFILING_LINKS) $(CLANG_LDFLAGS) $(EXPORT_LDFLAGS)
+AM_LDFLAGS = $(ZLIB_LINK) $(JSON_C_LINK) $(PW_CRACK_LINK) $(RUST_LDFLAGS) $(ASAN_CFLAGS) $(MSAN_CFLAGS) $(TSAN_CFLAGS) $(UBSAN_CFLAGS) $(PROFILING_LINKS) $(CLANG_LDFLAGS) $(EXPORT_LDFLAGS) $(FUZZ_FLAGS)
 endif #end hpux
 
 # https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html#Updating-version-info
@@ -846,7 +868,7 @@ librslapd_la_EXTRA = src/librslapd/Cargo.lock
 	SLAPD_DYLIB_DIR=$(abs_top_builddir)/ \
 	SLAPD_HEADER_DIR=$(abs_top_builddir)/ \
 		cargo rustc $(RUST_OFFLINE) --manifest-path=$(srcdir)/src/librslapd/Cargo.toml \
-		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS)
+		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS) $(RUSTC_SANITIZER_FLAGS)
 
 # The header needs the lib build first.
 rust-slapi-private.h: @abs_top_builddir@/rs/rslapd/@rust_target_dir@/librslapd.a
@@ -867,7 +889,7 @@ librnsslapd_la_EXTRA = src/librnsslapd/Cargo.lock
 	SLAPD_DYLIB_DIR=$(abs_top_builddir)/ \
 	SLAPD_HEADER_DIR=$(abs_top_builddir)/ \
 		cargo rustc $(RUST_OFFLINE) --manifest-path=$(srcdir)/src/librnsslapd/Cargo.toml \
-		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS)
+		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS)  $(RUSTC_SANITIZER_FLAGS)
 
 # The header needs the lib build first.
 rust-nsslapd-private.h: @abs_top_builddir@/rs/rnsslapd/@rust_target_dir@/librnsslapd.a
@@ -906,7 +928,7 @@ libentryuuid_la_EXTRA = src/plugin/entryuuid/Cargo.lock
 	SLAPD_DYLIB_DIR=$(abs_top_builddir)/ \
 	SLAPD_HEADER_DIR=$(abs_top_builddir)/ \
 		cargo rustc $(RUST_OFFLINE) --manifest-path=$(srcdir)/src/plugins/entryuuid/Cargo.toml \
-		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS)
+		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS) $(RUSTC_SANITIZER_FLAGS)
 	cp $(ENTRYUUID_LIB) @abs_top_builddir@/.libs/libentryuuid.a
 
 ENTRYUUID_SYNTAX_LIB = @abs_top_builddir@/rs/entryuuid_syntax/@rust_target_dir@/libentryuuid_syntax.a
@@ -924,7 +946,7 @@ libentryuuid_syntax_la_EXTRA = src/plugin/entryuuid_syntax/Cargo.lock
 	SLAPD_DYLIB_DIR=$(abs_top_builddir)/ \
 	SLAPD_HEADER_DIR=$(abs_top_builddir)/ \
 		cargo rustc $(RUST_OFFLINE) --manifest-path=$(srcdir)/src/plugins/entryuuid_syntax/Cargo.toml \
-		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS)
+		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS) $(RUSTC_SANITIZER_FLAGS)
 	cp $(ENTRYUUID_SYNTAX_LIB) @abs_top_builddir@/.libs/libentryuuid_syntax.a
 
 # == pwdchan
@@ -944,7 +966,7 @@ libpwdchan_la_EXTRA = src/plugin/pwdchan/Cargo.lock
 	SLAPD_DYLIB_DIR=$(abs_top_builddir)/ \
 	SLAPD_HEADER_DIR=$(abs_top_builddir)/ \
 		cargo rustc $(RUST_OFFLINE) --manifest-path=$(srcdir)/src/plugins/pwdchan/Cargo.toml \
-		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS)
+		$(CARGO_FLAGS) --verbose -- $(RUSTC_FLAGS) $(RUSTC_LINK_FLAGS) $(RUSTC_SANITIZER_FLAGS)
 	cp $(PWDCHAN_LIB) @abs_top_builddir@/.libs/libpwdchan.a
 
 # == pwdchan
@@ -1863,6 +1885,7 @@ ns_slapd_SOURCES = ldap/servers/slapd/abandon.c \
 	ldap/servers/slapd/house.c \
 	ldap/servers/slapd/init.c \
 	ldap/servers/slapd/main.c \
+	ldap/servers/slapd/fuzzer.c \
 	ldap/servers/slapd/monitor.c \
 	ldap/servers/slapd/passwd_extop.c \
 	ldap/servers/slapd/psearch.c \
@@ -1887,7 +1910,7 @@ ns_slapd_LDADD += $(LDAPSDK_LINK) $(NSS_LINK) $(LIBADD_DL) $(OPENSSL_LIBS) \
        $(NSPR_LINK) $(SASL_LINK) $(LIBNSL) $(LIBSOCKET) $(THREADLIB) $(SYSTEMD_LIBS) $(EVENT_LINK)
 
 ns_slapd_DEPENDENCIES = libslapd.la libldaputil.la
-ns_slapd_LINK = $(LINK)
+ns_slapd_LINK = $(CXXLINK) $(FUZZ_LINK_FLAGS)
 
 
 #------------------------
